File: CSharp/Selenium.WebTests/Pages/LoginPage.cs
using OpenQA.Selenium;
using Selenium.WebTests.Config;

namespace Selenium.WebTests.Pages;

public class LoginPage : BasePage
{
    private static readonly By UsernameInput = By.Id("user-name");
    private static readonly By PasswordInput = By.Id("password");
    private static readonly By LoginButton   = By.Id("login-button");
    private static readonly By ErrorMessage  = By.CssSelector("[data-test='error']");

    public LoginPage(IWebDriver driver) : base(driver) { }

    public void Navigate() => Driver.Navigate().GoToUrl(TestSettings.BaseUrl);

    public InventoryPage LoginAs(string username, string password)
    {
        WaitForElement(UsernameInput).SendKeys(username);
        Driver.FindElement(PasswordInput).SendKeys(password);
        WaitForClickable(LoginButton).Click();
        // Block until the inventory list renders so callers get a fully-loaded page.
        WaitForElement(By.ClassName("inventory_list"));
        return new InventoryPage(Driver);
    }

    public LoginPage LoginExpectingError(string username, string password)
    {
        WaitForElement(UsernameInput).Clear();
        WaitForElement(UsernameInput).SendKeys(username);
        Driver.FindElement(PasswordInput).Clear();
        Driver.FindElement(PasswordInput).SendKeys(password);
        WaitForClickable(LoginButton).Click();
        return this;
    }

    public bool HasError => IsElementVisible(ErrorMessage);
    public string ErrorText => HasError ? Driver.FindElement(ErrorMessage).Text : string.Empty;
}

File: Java/playwright-tests/src/main/java/com/qa/portfolio/pages/LoginPage.java
package com.qa.portfolio.pages;

import com.microsoft.playwright.Page;

public class LoginPage extends BasePage {

    private static final String BASE_URL      = "https://www.saucedemo.com";
    private static final String USERNAME_INPUT = "#user-name";
    private static final String PASSWORD_INPUT = "#password";
    private static final String LOGIN_BUTTON   = "#login-button";
    private static final String ERROR_MESSAGE  = "[data-test='error']";

    public LoginPage(Page page) {
        super(page);
    }

    public LoginPage navigate() {
        page.navigate(BASE_URL);
        return this;
    }

    public InventoryPage loginAs(String username, String password) {
        waitForVisible(USERNAME_INPUT);
        locate(USERNAME_INPUT).fill(username);
        locate(PASSWORD_INPUT).fill(password);
        locate(LOGIN_BUTTON).click();
        return new InventoryPage(page);
    }

    public LoginPage loginExpectingError(String username, String password) {
        waitForVisible(USERNAME_INPUT);
        locate(USERNAME_INPUT).fill(username);
        locate(PASSWORD_INPUT).fill(password);
        locate(LOGIN_BUTTON).click();
        return this;
    }

    public boolean hasError() {
        return locate(ERROR_MESSAGE).isVisible();
    }

    public String getErrorText() {
        return hasError() ? locate(ERROR_MESSAGE).textContent() : "";
    }
}
